### PR TITLE
Sequence: Specialise compute_sequence_values for builtin arithmetic types

### PR DESCRIPTION
**Pull request by bjude**
Only explicitly cast to T when we're dealing with builtin arithmetic types and just perform the `init + step * i` operation for all other types. This enables `thrust::sequence` to work for types without a conversion from `std::size_t`.

Addresses issue #1498

### CODE REVIEW

**Comment by GPUtester** [comment]
Can one of the admins verify this patch?

**Comment by allisonvacanti** [comment]
Thanks for the patch, and sorry about the review delay!

LGTM -- I just pushed a small fixup commit to add a missing newline. Starting tests now.

DVS CL: 30441766

run tests

**Comment by allisonvacanti** [comment]
There are some build failures to address before this is merged:

```
../testing/sequence.cu: In function 'void TestSequenceNoSizeTConversion()':
../testing/sequence.cu:152:20: error: comparison between signed and unsigned integer expressions [-Werror=sign-compare]
     for (auto i = 0; i < m.size(); ++i)
                      ~~^~~~~~~~~~
```

We also need to support C++11, so the `enable_if_t` will need to be changed to `enable_if`.

```
../thrust/system/detail/generic/sequence.inl:69:34: error: ‘enable_if_t’ is not a member of ‘std’
 struct compute_sequence_value<T, std::enable_if_t<std::is_arithmetic<T>::value>>
                                  ^~~
```

**Comment by allisonvacanti** [comment]
Thanks for the quick fix :) Restarting tests.

DVS CL: 30452736

run tests

File: testing/sequence.cu [hunk 1]
```diff
@@ -124,3 +124,36 @@ void TestSequenceComplex()
...
+
+void TestSequenceNoSizeTConversion()
+{
+    thrust::device_vector<Vector> m(64);
+    thrust::sequence(m.begin(), m.end(), ::Vector{0, 0}, ::Vector{1, 2});
+
+    for (std::size_t i = 0; i < m.size(); ++i)
+    {
+        const ::Vector v = m[i];
+        ASSERT_EQUAL(v.x, i);
```
Comment (allisonvacanti) [inline testing/sequence.cu:155]: Looks like this needs a couple more cast to build cleanly:

```
../testing/unittest/assertions.h:121:12: error: comparison of integers of different signs: 'int' and 'unsigned long' [-Werror,-Wsign-compare]
    if(!(a == b)){
         ~ ^  ~
/jenkins/workspace/nvidia/thrust/prb/thrust-cpu-build/CXX_TYPE/clang/CXX_VER/10/OS_TYPE/ubuntu/OS_VER/20.04/SDK_TYPE/cuda/SDK_VER/11.3.1-devel/testing/sequence.cu:155:9: note: in instantiation of function template specialization 'unittest::assert_equal<int, unsigned long>' requested here
        ASSERT_EQUAL(v.x, i);
        ^
../testing/unittest/assertions.h:70:37: note: expanded from macro 'ASSERT_EQUAL'
#define ASSERT_EQUAL(X,Y)           ASSERT_EQUAL_WITH_FILE_AND_LINE((X),(Y), __FILE__,  __LINE__)
                                    ^
../testing/unittest/assertions.h:13:78: note: expanded from macro 'ASSERT_EQUAL_WITH_FILE_AND_LINE'
#define ASSERT_EQUAL_WITH_FILE_AND_LINE(X,Y,FILE_,LINE_)           unittest::assert_equal((X),(Y), FILE_,  LINE_)
```

**Review by allisonvacanti — CHANGES_REQUESTED** [review]

**Comment by bjude** [comment]
Damn, that one didnt show up with MSVC, they seem to be a bit inconsistent with when the signed vs unsigned warning triggers.

I'll get it fixed up today

**Comment by allisonvacanti** [comment]
Thanks!

DVS CL: 30482082

run tests

**Pull request merged**